File: Questions/547. Number of Provinces/main.cpp
#include<iostream>
using namespace std;

vector<vector<int>> g;
vector<bool> visited;
int ans;

void dfs(int idx) {
    if (visited[idx]) return;
    visited[idx] = true;
    for (int j : g[idx]) {
        if (!visited[j]) {
            dfs(j);
            ans++;
        }
    }
}

int findCircleNum(vector<vector<int>>& isConnected) {
    int n = isConnected.size();
    vector<vector<int>> g;
    visited.assign(false, n);

    for(int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            if (isConnected[i][j]) {
                g[i].push_back(j);
                g[j].push_back(i);
            }
        }
    }
    
    for (int i = 0; i < n; i++) {
        dfs(i);
    }
    return ans;
}

File: Questions/547. Number of Provinces/main.py
from collections import defaultdict
def find_circle_num(is_connected: list[list[int]]) -> int:
    g = defaultdict(list)
    n = len(is_connected)

    for i in range(n):
        for j in range(n):
            if is_connected[i][j]:
                g[i].append(j)
                g[j].append(i)

    visited = [False] * n

    def dfs(i: int) -> None:
        if visited[i]:
            return
        visited[i] = True
        for j in g[i]:
            if not visited[j]:
                dfs(j)
    ans = 0
    for i in range(n):
        if not visited[i]:
            dfs(i)
            ans += 1
    return ans

print(find_circle_num([[1,1,0],[1,1,0],[0,0,1]]))
